src/openzaak/utils/tasks.py

```diff
@@ -0,0 +1,791 @@
+
+    file_path = instance.report_file.file.name if instance.report_file else None
+    file_exists = Path(instance.report_file.file.name).exists() if file_path else False
+    has_data = bool(_get_total_count(file_path)) if file_exists else False
+
+    if file_exists and has_data:
+        mode = "a"
+    elif file_exists:
+        mode = "w"
+    else:
+        mode = "w+"
+
+    logger.debug(f"Using file mode {mode} for file {file_path or default_path}")
+
+    with open(file_path or default_path, mode) as _export_file:
+        csv_writer = csv.writer(_export_file, delimiter=",", quotechar='"')
+
+        if mode in ("w", "w+"):
+            csv_writer.writerow(DocumentRow.export_headers)
+
+        for row in batch:
+            data = row.as_export_data()
+            csv_writer.writerow(data.values())
+
+    if file_path:
+        return
+
+    relative_path = Path(instance.report_file.field.upload_to) / default_name
+
+    instance.report_file.name = str(relative_path)
+
+    try:
+        instance.save(update_fields=["report_file"])
+    except DatabaseError as e:
+        logger.critical(
+            f"Unable to save new report file due to database error: {str(e)}"
+        )
+
+
+def _finish_import(
+    instance: Import,
+    status: ImportStatusChoices,
+    finished_on: Optional[datetime] = None,
+    comment: Optional[str] = "",
+):
+    updated_fields = ["finished_on", "status"]
+
+    instance.finished_on = finished_on or timezone.now()
+    instance.status = status
+
+    if comment:
+        instance.comment = comment
+
+        updated_fields.append("comment")
+
+    logger.info(f"Finishing import with status {status.label}")
+
+    try:
+        instance.save(update_fields=updated_fields)
+    except DatabaseError as e:
+        logger.critical(f"Unable to save import state due to database error: {str(e)}")
+
+
+def _finish_batch(import_instance: Import, batch: list[DocumentRow],) -> None:
+    batch_number = import_instance.get_batch_number(len(batch))
+    _processed, _fail_count, _success_count = _get_batch_statistics(batch)
+
+    import_instance.processed = import_instance.processed + _processed
+    import_instance.processed_successfully = (
+        import_instance.processed_successfully + _success_count
+    )
+    import_instance.processed_invalid = import_instance.processed_invalid + _fail_count
+
+    try:
+        import_instance.save(
+            update_fields=["processed", "processed_successfully", "processed_invalid"]
+        )
+    except DatabaseError as e:
+        logger.critical(
+            f"Unable to save batch statistics for batch {batch_number} due to database "
+            f"error: {str(e)}"
+        )
+
+    logger.info(f"Writing batch number {batch_number} to report file")
+    _write_to_file(import_instance, batch)
+
+
+LOCK_EXPIRE = 60 * (60 * 24)  # 24 hours
+
+
+@contextmanager
+def task_lock(lock_id, oid):
+    timeout_at = monotonic() + LOCK_EXPIRE - 3
```

Comment:
Is there a specific reason we need to use `monotonic` here? Also why is the `- 3` needed? It looks a bit random :stuck_out_tongue: